### PR TITLE
[WIP] Initialize all members with in-class field initializers

### PR DESCRIPTION
**Pull request by akien-mga**
**IMPORTANT:** This is a WIP and should not be merged as is, it breaks the build and likely does some bogus initializations.

These changes are the result of applying `clang-tidy`'s [`cppcoreguidelines-pro-type-member-init`](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines-pro-type-member-init.html) check to the codebase, with the aim to have all our class fields properly initialized to sane defaults (`0`, `false`, `nullptr`).

`clang-tidy` made a huge mess of the codebase so I had to fix quite a few things manually, such as:
```diff
--- a/core/variant/variant.h
+++ b/core/variant/variant.h
@@ -204,7 +204,7 @@ private:
                PackedArrayRefBase *packed_array;
                void *_ptr; //generic pointer
                uint8_t _mem[sizeof(ObjData) > (sizeof(real_t) * 4) ? sizeof(ObjData) : (sizeof(real_t) * 4)];
-       } _data alignas(8);
+       } _data{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{}{} alignas(8);
 
        void reference(const Variant &p_variant);
 
```
```diff
--- a/core/math/color.h
+++ b/core/math/color.h
@@ -37,10 +37,10 @@
 struct Color {
        union {
                struct {
-                       float r;
-                       float g;
-                       float b;
-                       float a;
+                       float r = 0.0F = 0.0F = 0.0F = 0.0F;
+                       float g = 0.0F = 0.0F = 0.0F = 0.0F;
+                       float b = 0.0F = 0.0F = 0.0F = 0.0F;
+                       float a = 0.0F = 0.0F;
                };
                float components[4] = { 0, 0, 0, 1.0 };
        };
```

It also seems to be completely clueless around unions and introduces bugs by initializing members with incompatible values... So there's **a lot** of post-processing work needed to make this mess actually useful, to the point that it might not be useful at all as a starting point... But let's review some of the changes and see if they can be salvaged or redone manually.

### CODE REVIEW

File: core/debugger/local_debugger.cpp [hunk 1]
```diff
@@ -84,7 +84,7 @@ struct LocalDebugger::ScriptsProfiler {
 			}
 		}
 
-		SortArray<ScriptLanguage::ProfilingInfo, ProfileInfoSort> sort;
+		SortArray<ScriptLanguage::ProfilingInfo, ProfileInfoSort> sort{};
```
Comment (akien-mga) [inline core/debugger/local_debugger.cpp:87]: I'm not sure why it wants to initialize `SortArray` like this but not `Vector` and `TypedArray`. The only field in `SortArray` seems to be the `Comparator` template instance?

File: core/input/input.cpp [hunk 2]
```diff
@@ -654,7 +654,7 @@ void Input::start_joy_vibration(int p_device, float p_weak_magnitude, float p_st
 	if (p_weak_magnitude < 0.f || p_weak_magnitude > 1.f || p_strong_magnitude < 0.f || p_strong_magnitude > 1.f) {
 		return;
 	}
-	VibrationInfo vibration;
+	VibrationInfo vibration{};
```
Comment (akien-mga) [inline core/input/input.cpp:657]: Maybe this `struct` can be initialized in its declaration instead of when it's used?

File: core/input/input.h [hunk 3]
```diff
@@ -130,10 +130,10 @@ class Input : public Object {
...
 
 	struct SpeedTrack {
-		uint64_t last_tick;
+		uint64_t last_tick = 0;
 		Vector2 speed;
 		Vector2 accum;
-		float accum_t;
+		float accum_t = 0.0;
 		float min_ref_frame;
 		float max_ref_frame;
```
Comment (akien-mga) [inline core/input/input.h:138]: These two are initialized in constructor, this should be moved here.

File: core/math/color.h [hunk 4]
```diff
@@ -37,10 +37,10 @@
...
 	union {
 		struct {
-			float r;
-			float g;
-			float b;
-			float a;
+			float r = 0.0;
+			float g = 0.0;
+			float b = 0.0;
+			float a = 0.0;
```
Comment (akien-mga) [inline core/math/color.h:43]: `clang-tidy` obviously doesn't have a clue about what it's doing with unions...

I guess the best short-term fix is to undo all the changes it's trying to do in unions.

Is there a "proper" way to in-class initialize a union, especially with non-zero data like here? (poke @hpvb)

File: core/math/dynamic_bvh.h [hunk 5]
```diff
@@ -182,7 +182,7 @@ class DynamicBVH {
 		Volume volume;
 		Node *parent = nullptr;
 		union {
-			Node *childs[2];
+			Node *childs[2]{};
```
Comment (akien-mga) [inline core/math/dynamic_bvh.h:185]: The constructor can be removed if this is a correct way to initialize both elements to `nullptr`.
```
		Node() {
			childs[0] = nullptr;
			childs[1] = nullptr;
		}
```

File: core/variant/variant.h [hunk 6]
```diff
@@ -204,7 +204,9 @@ class Variant {
 		PackedArrayRefBase *packed_array;
 		void *_ptr; //generic pointer
 		uint8_t _mem[sizeof(ObjData) > (sizeof(real_t) * 4) ? sizeof(ObjData) : (sizeof(real_t) * 4)];
-	} _data alignas(8);
+	} _data {
+	}
+	alignas(8);
```
Comment (akien-mga) [inline core/variant/variant.h:209]: Aside from the bogus style changes from `clang-format`, I assume that this is *not* something that we want to do? I'm not even sure how that would work, I'm having trouble finding any proper documentation on how unions should or can be initialized.

**Comment by qarmin** [comment]
`cppcoreguidelines-prefer-member-initializer` - may be added also, but it will be available in llvm 13
https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines-prefer-member-initializer.html

File: drivers/unix/dir_access_unix.cpp [hunk 7]
```diff
@@ -77,7 +77,7 @@ bool DirAccessUnix::file_exists(String p_file) {
 
 	p_file = fix_path(p_file);
 
-	struct stat flags;
+	struct stat flags {};
```
Comment (akien-mga) [inline drivers/unix/dir_access_unix.cpp:80]: I don't know why `clang-format` enforces a space here when it doesn't bother about other types of initializers. I guess it's confused by the `struct` part?

**Comment by akien-mga** [comment]
I split the `core` changes to #47748 and did further fixes there. Let's focus on `core` first, decide what to do for unions, etc., and the apply the same logic to the rest of the codebase.

**Pull request closed without merging**